File: TwoPointer/766.toeplitz-matrix.go
package leetcode

func isToeplitzMatrix(matrix [][]int) bool {
	nr, nc := len(matrix), len(matrix[0])
	for i := 1; i < nr; i++ {
		for j := 1; j < nc; j++ {
			if matrix[i][j] != matrix[i-1][j-1] {
				return false
			}
		}
	}
	return true
}


// @leet start
func isToeplitzMatrix(matrix [][]int) bool {
	nrow, ncol := len(matrix), len(matrix[0])
	check := func(r, c int) bool {
		num := matrix[r][c]
		for r < nrow && c < ncol {
			if matrix[r][c] != num {
				return false
			}
			r++
			c++
		}
		return true
	}
	for r := range nrow {
		if !check(r, 0) {
			return false
		}
	}
	for c := range ncol {
		if !check(0, c) {
			return false
		}
	}
	return true
}

// @leet end

File: TwoPointer/766.toeplitz-matrix.py
from typing import List


# @leet start
class Solution:
    def isToeplitzMatrix(self, matrix: List[List[int]]) -> bool:
        nr, nc = len(matrix), len(matrix[0])

        def check(r, c):
            n = matrix[r][c]
            while r < nr and c < nc:
                if matrix[r][c] != n:
                    return False
                r += 1
                c += 1
            return True

        for r in range(nr):
            if not check(r, 0):
                return False
        for c in range(nc):
            if not check(0, c):
                return False
        return True


# @leet end

